src/Coordination/KeeperServer.cpp

```diff
@@ -906,4 +907,22 @@ Keeper4LWInfo KeeperServer::getPartiallyFilled4LWInfo() const
     return result;
 }
 
+bool KeeperServer::createSnapshot()
+{
+    std::lock_guard lock(snapshot_mutex);
+    if (raft_instance->create_snapshot())
+    {
+        last_manual_snapshot_log_idx = raft_instance->get_last_snapshot_idx();
```

Comment:
```suggestion
        last_manual_snapshot_log_idx = raft_instance->get_last_snapshot_idx();
```
I think this line needs to be removed.
You instantly set `last_manual_snapshot_log_idx` to `get_last_snapshot_idx` and then your condition for done is if they are the same value.
It needs to be set after the scheduled task is done IMO.

Instead of having `snpd`, what about adding a more general 4LW which will contain some info about the logs.
E.g. `last_created_snapshot_idx`, `last_uploaded_snapshot_idx`, `last_committed_idx`...?